File: apps/vsc-plugin/src/services/agent-detection/detectors/coderabbit.detector.ts
import { findExtension } from '../checks/extension';
import { whichBinary } from '../checks/binary';
import { dirExists, expandHome } from '../checks/config-dir';
import type { AgentDetector, DetectionContext, DetectionResult } from '../types';

/**
 * Detects CodeRabbit (CR's agentic PR-review CLI). Three signals
 * checked in order — first hit wins:
 *
 *   1. The official VS Code extension (`coderabbitai.coderabbit-vscode`).
 *   2. The `coderabbit` binary on PATH (installed via
 *      `curl -fsSL https://cli.coderabbit.ai/install.sh | sh`).
 *   3. The `~/.coderabbit/` config directory (created after a
 *      successful `coderabbit login`).
 *
 * Every branch returns `isTerminalAgent: true`. The plugin does not
 * drive CodeRabbit directly; codeam-cli's BatchAgentStrategy handles
 * runtime end-to-end.
 */
export class CodeRabbitDetector implements AgentDetector {
  readonly id = 'coderabbit';
  readonly name = 'CodeRabbit';
  readonly icon = 'coderabbit';

  private static readonly CANDIDATE_EXTENSION_IDS = ['coderabbitai.coderabbit-vscode'];
  private static readonly BINARY_NAME = 'coderabbit';
  private static readonly CONFIG_DIR = '~/.coderabbit';

  async detect(ctx: DetectionContext): Promise<DetectionResult | null> {
    const extension = findExtension(CodeRabbitDetector.CANDIDATE_EXTENSION_IDS, ctx.extensions);
    if (extension) {
      return {
        installed: true,
        extensionId: extension.id,
        isTerminalAgent: true,
        via: 'extension',
      };
    }

    const binPath = await whichBinary(CodeRabbitDetector.BINARY_NAME);
    if (binPath) {
      ctx.log.appendLine(`[CodeRabbitDetector] binary at ${binPath}`);
      return {
        installed: true,
        extensionId: `__binary__:${CodeRabbitDetector.BINARY_NAME}`,
        isTerminalAgent: true,
        via: 'binary',
      };
    }

    if (await dirExists(expandHome(CodeRabbitDetector.CONFIG_DIR))) {
      ctx.log.appendLine(`[CodeRabbitDetector] config dir ${CodeRabbitDetector.CONFIG_DIR} present`);
      return {
        installed: true,
        extensionId: `__config__:${CodeRabbitDetector.BINARY_NAME}`,
        isTerminalAgent: true,
        via: 'config-dir',
      };
    }

    return null;
  }
}

File: apps/jetbrains-plugin/src/main/kotlin/com/windsurf/controller/services/detection/detectors/CodeRabbitDetector.kt
package com.windsurf.controller.services.detection.detectors

import com.windsurf.controller.services.detection.AgentDetector
import com.windsurf.controller.services.detection.DetectionContext
import com.windsurf.controller.services.detection.DetectionResult
import com.windsurf.controller.services.detection.checks.PluginRef
import com.windsurf.controller.services.detection.checks.dirExists
import com.windsurf.controller.services.detection.checks.expandHome
import com.windsurf.controller.services.detection.checks.findInstalledPlugin
import com.windsurf.controller.services.detection.checks.whichBinary

/**
 * CodeRabbit — terminal agent owned by codeam-cli's BatchAgentStrategy.
 * Probes the marketplace plugin (id best-effort), the `coderabbit`
 * binary, and the `~/.coderabbit/` config directory.
 */
class CodeRabbitDetector : AgentDetector {
    override val id = "coderabbit"
    override val name = "CodeRabbit"
    override val icon = "coderabbit"

    private val candidatePlugins = listOf(
        PluginRef(id = "com.coderabbit.coderabbit-jetbrains"),
    )
    private val binaryName = "coderabbit"
    private val configDir = "~/.coderabbit"

    override suspend fun detect(ctx: DetectionContext): List<DetectionResult> {
        val plugin = findInstalledPlugin(candidatePlugins)
        val installed = plugin != null ||
            whichBinary(binaryName) != null ||
            dirExists(expandHome(configDir))
        if (!installed) return emptyList()
        ctx.logger.info("[CodeRabbitDetector] detected (plugin=${plugin?.id ?: "none"})")
        return listOf(
            DetectionResult(
                id = id,
                name = name,
                icon = icon,
                pluginId = plugin?.id ?: "coderabbit.coderabbit",
                toolWindowId = "__terminal__:$id",
                isTerminalAgent = true,
            ),
        )
    }
}
